clippy_lints/src/unused_peekable.rs

```diff
@@ -0,0 +1,151 @@
+                    if let Some(expr) = block.expr {
+                        vis.visit_expr(expr);
+                    } else {
+                        return;
+                    }
+                } else {
+                    for stmt in &block.stmts[idx..] {
+                        vis.visit_stmt(stmt);
+                    }
+                }
+
+                if !vis.found_peek_call {
+                    span_lint_and_help(
+                        cx,
+                    UNUSED_PEEKABLE,
+                        ident.span,
+                        "`peek` never called on `Peekable` iterator",
+                        None,
+                        "consider removing the call to `peekable`"
+                   );
+                }
+            }
+        }
+    }
+}
+
+struct PeekableVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    expected_ident: &'a str,
+    found_peek_call: bool,
+}
+
+impl<'a, 'tcx> PeekableVisitor<'a, 'tcx> {
+    fn new(cx: &'a LateContext<'tcx>, expected_ident: &'a str) -> Self {
+        Self {
+            cx,
+            expected_ident,
+            found_peek_call: false,
+        }
+    }
+}
+
+impl<'tcx> Visitor<'_> for PeekableVisitor<'_, 'tcx> {
+    fn visit_expr(&mut self, ex: &'_ Expr<'_>) {
+        match ex.kind {
+            // some_function(peekable)
+            //
+            // If the Peekable is passed to a function, stop
+            ExprKind::Call(_, args) => {
+                for arg in args.iter().map(|arg| peel_ref_operators(self.cx, arg)) {
+                    if let ExprKind::Path(ref qpath) = arg.kind
+                        && match_qpath(qpath, &[self.expected_ident])
+                        && let Some(ty) = self.cx.typeck_results().expr_ty_opt(arg).map(Ty::peel_refs)
+                        && match_type(self.cx, ty.peel_refs(), &paths::PEEKABLE)
+                    {
+                        self.found_peek_call = true;
+                        return;
+                    }
+                }
+            },
+            // Peekable::peek()
+            ExprKind::MethodCall(PathSegment { ident: method_name, .. }, [arg, ..], _) => {
+                let arg = peel_ref_operators(self.cx, arg);
+                let method_name = method_name.name.as_str();
+
+                if (method_name == "peek"
+                    || method_name == "peek_mut"
+                    || method_name == "next_if"
+                    || method_name == "next_if_eq")
+                    && let ExprKind::Path(ref qpath) = arg.kind
+                    && match_qpath(qpath, &[self.expected_ident])
+                    && let Some(ty) = self.cx.typeck_results().expr_ty_opt(arg).map(Ty::peel_refs)
+                    && match_type(self.cx, ty, &paths::PEEKABLE)
+                {
+                    self.found_peek_call = true;
+                    return;
+                }
+            },
+            _ => {},
+        }
+
+        walk_expr(self, ex);
+    }
+}
```

Comment:
This is a tricky one, there's a lot of ways a peekable could end up being used and I'm sure we'll miss some extra ones on top of that

e.g.

```rust
let mut lexer = Lexer { tokens: peekable };
lexer.tokens.peek();

let mut new_binding = peekable;
new_binding.peek();

let by_ref = &mut peekable;
by_ref.peek();
```

A potential way to handle this would be in the visitor check if `ex` is a path to the peekable binding, then use `parent_iter` to check any potential uses it has, considering it used if we're unsure